src/rgw/rgw_lc.cc

```diff
@@ -2400,14 +2400,19 @@ static int guard_lc_modify(const DoutPrefixProvider *dpp,
   utime_t time(max_lock_secs, 0);
 
   int ret;
+  uint16_t retries{0};
 
+  // due to reports of starvation trying to save lifecycle policy, try hard
   do {
     ret = lock->try_lock(dpp, time, null_yield);
     if (ret == -EBUSY || ret == -EEXIST) {
       ldpp_dout(dpp, 0) << "RGWLC::RGWPutLC() failed to acquire lock on "
-          << oid << ", sleep 5, try again" << dendl;
-      sleep(5); // XXX: return retryable error
-      continue;
+			<< oid << ", retry in 100ms, ret=" << ret << dendl;
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+      // the typical S3 client will time out in 60s
+      if(retries++ < 500) {
+	continue;
+      }
```

Comment:
we'll eventually want to avoid these blocking sleeps, because they're called from `RGWPutLC` under a coroutine from the frontend. if we want to keep this polling model, we should add support for `optional_yield` with something like `RGWReshardWait::wait(optional_yield y)`

could it make sense to 'stage' these policy updates somewhere that doesn't require a lock? such that, once lifecycle starts processing the next 'day', the first worker to get the lock would copy any 'staged' updates into the LC list

Reply:
wouldn't that workflow need to serialize the append-work step (albeit, it could content for a different lock)?

Reply:
optional yield stuff noted, can do that